scarb/src/bin/scarb/commands/run.rs

```diff
@@ -7,38 +7,60 @@ use indoc::formatdoc;
 use serde::{Serialize, Serializer};
 use smol_str::SmolStr;
 
+use scarb::core::errors::ScriptExecutionError;
 use scarb::core::{Config, Package, Workspace};
 use scarb::ops;
 use scarb::ui::Message;
 
 use crate::args::ScriptsRunnerArgs;
+use crate::errors::ErrorWithExitCode;
 
 #[tracing::instrument(skip_all, level = "info")]
 pub fn run(args: ScriptsRunnerArgs, config: &Config) -> Result<()> {
     let ws = ops::read_workspace(config.manifest_path(), config)?;
-    let package = args.packages_filter.match_one(&ws)?;
-    if let Some(script) = args.script {
-        run_script(script, args.args, package, &ws)
+    let packages = args.packages_filter.match_many(&ws)?;
+    let errors = packages
+        .into_iter()
+        .map(|package| {
+            args.script
+                .clone()
+                .map(|script| run_script(script, &args.args, package.clone(), &ws))
+                .unwrap_or_else(|| list_scripts(package, config))
+        })
+        .filter_map(|res| res.err())
+        .map(|res| anyhow!(res))
+        .collect::<Vec<anyhow::Error>>();
+    if errors.is_empty() {
+        Ok(())
     } else {
-        list_scripts(package, config)
+        let exit_code = errors
+            .iter()
+            .filter_map(|err| {
+                err.downcast_ref::<ScriptExecutionError>()
+                    .map(|err| err.exit_code)
+            })
+            .find(|exit_code| *exit_code != 0)
+            .unwrap_or(1);
+        let msg = errors
+            .into_iter()
+            .map(|err| err.to_string())
+            .collect::<Vec<String>>()
+            .join("\n");
+        Err(ErrorWithExitCode::new(anyhow!(msg), exit_code).into())
     }
 }
 
-fn run_script(
-    script: SmolStr,
-    args: Vec<OsString>,
-    package: Package,
-    ws: &Workspace,
-) -> Result<()> {
+fn run_script(script: SmolStr, args: &[OsString], package: Package, ws: &Workspace) -> Result<()> {
     let script_definition = package.manifest.scripts.get(&script).ok_or_else(|| {
+        let package_name = package.id.name.to_string();
         anyhow!(formatdoc! {r#"
-            missing script `{script}`
+            missing script `{script}` for package `{package_name}`
 
             To see a list of scripts, run:
-                scarb run
+                scarb run -p {package_name}
```

Comment:
If you have single-package workspace, I would rather skip printing this argument